### DISCUSSION TITLE
Walking Direction Detection with Frigate and NodeRed

### DISCUSSION THREAD

**Discussion by xury77** [blakeblackshear/frigate#15460] [Ask A Question] [answered]
I would like to use events and NodeRed to detect the direction of a person moving from zone one to zone two in order to be able to execute commands to open a small pedestrian gate. A necessary condition is that the command is executed only when the movement is from zone one to zone two. Below is my code fragment for the function node, but I don't understand why it doesn't work well.

![Zrzut ekranu z 2024-12-12 11-03-13](https://github.com/user-attachments/assets/e1cc1bfb-3f5a-4500-8bb1-1fc156ea6ee3)

```
let beforeZones = msg.payload.before?.current_zones || [];

let afterZones = msg.payload.after?.current_zones || [];
let label = msg.payload.after?.label || "";

// Check if the object is "person" and has moved from "zone_one" to "zone_two"
if (label === "person" && beforeZones.includes("zone_one") && afterZones.includes("zone_two")) {
    msg.payload = {
        action: "open_gate",
        timestamp: msg.payload.after.frame_time, 
        zones: {
            from: "zone_one",
            to: "zone_two"
        }
    };
    return msg;
}

return null;
```

**Comment by NickM-27** [comment]
You need to have a check to ensure it only fires the first time the zones change

**Comment by xury77** [comment]
Yes, I will do it later. Currently I do not get any message on the output of the function node. I think that the zones do not have to slightly overlap to meet the condition: `msg.payload.before.current_zones["zone_one"]`
with:
`msg.payload.after.current_zones["zone_two"]`
I have to check this.

**Reply by jkoopmann** [reply]
have you gotten this to work @xury77 ?

I wanted to to something very similar (only direction of a car moving through a camera feed). Also frigate, mqtt and nodered. It would be brilliant if you could share your solution.

**Reply by xury77** [reply, marked as answer]
Almost same flow working for me now:

```
// Sprawdź obecność klucza i typ obiektu
let beforeZones = msg.payload.before?.current_zones || [];
let afterZones = msg.payload.after?.current_zones || [];

// Sprawdź, czy to zdarzenie dotyczy obiektu "person"
if (msg.payload.after?.label === "person") {
    // Sprawdź, czy obiekt przeszedł z "wjazd" do "dojscie"
    if (beforeZones.includes("wjazd") && afterZones.includes("dojscie")) {
        msg.topic = 263;
        msg.payload = true;
        return msg;
    }
}

// Zwróć wiadomość
return null;
```
